src/commandargs.cpp

```diff
@@ -23,6 +34,8 @@ void ParseCommandArgs(int argc, char* argv[], ProgramConfiguration config)
 {
     Config = &config;
     CurrentState = nullptr;
+    bool inFlag = false;
+    programName = "./program.pebl";
```

Comment:
Commandargs is intended to be relatively generalizable (so not bound to running just pebble from the commandline). Can we make programName a Setting in the ProgramConfiguration of Pebble (see `main.cpp`)

Reply:
We could make programName a Setting, for sure. 

I think a way of declaring a usage scheme for the flags is the most valuable improvement to be made for a commandline parsing library. 

Some examples I personally have enjoyed for interfaces in other languages (upside and downsides present):

https://github.com/henesy/bankfs/blob/master/main.c
```c
void
main(int argc, char *argv[])
{
	char	*mnt, *srv, *addr, *ndb, *hostuser, *hostgrp;

	ndb			=	nil;
	srv			=	nil;
	hostuser	=	"glenda";
	hostgrp		=	"sys";
	mnt			=	"/mnt/bankfs";
	// Disable for development, listener works
	addr		=	nil;
	//addr		=	"tcp!*!3656";

	ARGBEGIN{
	case 'D':
		// Debug chatter
		chatty9p++;
		break;
	case 's':
		// #s name to post
		srv = EARGF(usage());
		break;
	case 'm':
		// Directory path to mount to
		mnt = EARGF(usage());
		break;
	case 'a':
		// Dial string address to listen on
		addr = EARGF(usage());
		break;
	case 'd':
		// .ndb file to load from (if any)
		ndb = EARGF(usage());
		break;
	case 'u':
		// User to root as
		hostuser = EARGF(usage());
		break;
	case 'g':
		// Group to root as
		hostgrp = EARGF(usage());
		break;
	default:
		usage();
	}ARGEND;

	if(argc)
		usage();
```

https://github.com/henesy/limbobyexample/blob/master/Args-M/args.b
```c
init(nil: ref Draw->Context, argv: list of string) {
	sys = load Sys Sys->PATH;
	arg := load Arg Arg->PATH;
	if(arg == nil)
		raise "bad module load";

	arg->init(argv);
	arg->setusage(arg->progname() + " [-r] [-c mark] words...");

	rev := 0;
	mark := "→ ";

	while((opt := arg->opt()) != 0)
		case opt {
		'r' =>
			rev = 1;
		'c' =>
			mark = arg->earg();
		* =>
			arg->usage();
		}

	argv = arg->argv();

	print("Argc after flags: %d\n", len argv);

	…

	for(; argv != nil; argv = tl argv)
		sys->print("%s %s\n", mark, hd argv);
```

https://github.com/henesy/socketh-myr/blob/master/server/main.myr
```ml
const main = {args : byte[:][:]
	var cmd
	var addr = "tcp!*!9090"
	sleeptime = 5_000_000
	showprompt = false

	// Commandline argument parsing
	cmd = std.optparse(args, &[
	    .argdesc = "",
	    .opts = [
	        [.opt='D',						.desc = "enable verbose logging"],
	        [.opt='P',						.desc = "enable prompt display"],
	        [.opt='p',	.arg = "prompt",	.desc = "set prompt to 'prompt'"],
	        [.opt='T',	.arg = "period",	.desc = "set sleep time to 'period' ns"],
	        [.opt='a',	.arg = "address",		.desc = "listen on dial string 'address'"],
	    ][:]
	])

	for opt : cmd.opts
	    match opt
	    | ('D', ""):
	    	chatty = true
	    | ('P', ""):
	    	showprompt = true
	    | ('p', arg):
	    	prompt = arg
	    | ('T', arg):
			match std.intparse(arg)
			| `std.Some v:
	    		sleeptime = v
			| `std.None:
				std.fatal("err: invalid argument for sleeptime")
			;;
	    | ('a', arg):
	    	addr = arg
	    | _:
	    	std.fatal("unreachable\n")
	    ;;
	;;
```

https://github.com/henesy/SocketS/blob/master/server/main.go
```go
func main() {
	flag.StringVar(&port, "p", ":9090", "Port to listen on")
	flag.Uint64Var(&maxConns, "m", 100, "Maximum chat connections")
	flag.UintVar(&maxMsgSize, "s", 256, "Max size of a message read out")
	flag.UintVar(&maxUsrSize, "u", 25, "Max size of username")
	flag.Parse()
```

Note that Go's flag package has a way to extract the remaining arguments _after_ flag parsing (very ergonomic): https://golang.org/pkg/flag/#Args

Most of the other libraries shown here do as well, Limbo's is via `arg.argv()`: http://man.cat-v.org/inferno/2/arg

For plan9's [arg(2)](http://man.cat-v.org/plan_9/2/arg) `argv` is modified in place and `argv0` is a special case where `argv[0]`'s original string is preserved. 

For Myrddin the standard library parser exposes the remaining commandline arguments in the parsed type form of the commandline options structure: https://myrlang.org/doc/libstd/cli.html